Updates to taurus initialization script

oxtopus opened the pull request:
This initializes a pair of instances to the point of being able to run tests!  cc @jcasner

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
+"
+
+if [ "${1}" == "-h" ]; then
+  echo "$USAGE"
```
s/`"$USAGE"`/`"${USAGE}"`

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
 set -o pipefail
-set -o verbose
 set -o nounset
+set -o verbose
```
Considering using a `DEBUG` flag - either as an env var or a command line param - to turn on `verbose` mode

oxtopus replied:
I originally felt compelled to force this as verbose, but if that's not desired, I think it's best to remove the line altogether, not add any special flags, and let the user invoke `bash -v` externally if they want it verbose.  For example, instead of `./initialize-remote-taurus-servers.sh`, the verbose version would be `/bin/bash -v ./initialize-remote-taurus-servers.sh`

jcasner replied:
I'll defer to @unixorn on this one...

unixorn replied:
I'd prefer checking if `${DEBUG}` is set to calling a script with `bash -v`.

It feels non-unixy to do interpreter script instead of having a proper shebang line.

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
+  # Generate env.sh credential files, to be copied to respective remote
+  # instances later
+  echo "
+    export PATH=/opt/numenta/anaconda/bin:\$PATH
```
I don't think you need the `\` before the `$` here.  I think it should be s/`\$PATH``/${PATH}`

oxtopus replied:
The $PATH and $PYTHONPATH need to be escaped in the remote environment.  Without the escapes, the local values are substituted during execution, which is not what we want.  This is deliberately different from the other vars in this set.

jcasner replied:
Aha, understood.  That makes sense!

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
+  # instances later
+  echo "
+    export PATH=/opt/numenta/anaconda/bin:\$PATH
+    export PYTHONPATH=/opt/numenta/anaconda/lib/python2.7/site-packages:\$PYTHONPATH
```
s/`\$PYTHONPATH`/`${PYTHONPATH}`

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
 
-  # Reset server state, apply database schema, start taurus services
+  echo "
+    export PATH=/opt/numenta/anaconda/bin:\$PATH
```
s/`\$PATH`/`${PATH}`

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
-  # Reset server state, apply database schema, start taurus services
+  echo "
+    export PATH=/opt/numenta/anaconda/bin:\$PATH
+    export PYTHONPATH=/opt/numenta/anaconda/lib/python2.7/site-packages:\$PYTHONPATH
```
s/`\$PYTHONPATH`/`${PYTHONPATH}`

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
+        /opt/numenta/anaconda/lib/python2.7/site-packages \
+        /opt/numenta/anaconda/bin &&
+     taurus-set-rabbitmq \
+        --host=127.0.0.1 \
```
Should we also use a `RABBITMQ_HOST` as we've done for MySQL?  In Staging and Production, for example, we use a remote server...

oxtopus replied:
I tried that, but it didn't work.  Maybe something about the routing and dns.  On that note, we do need to do something, since we won't always be running rabbitmq on the same host.

jcasner replied:
It's interesting that RABBITMQ_HOST works for the collector...

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
+    "${TAURUS_COLLECTOR_USER}"@"${TAURUS_COLLECTOR_HOST}":/opt/numenta/products \
+    `git rev-parse --abbrev-ref HEAD`
+
+  # Reset server state
```
s/`server`/`collector`

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
-     cd /opt/numenta/products/taurus.metric_collectors/
-     supervisorctl -s http://127.0.0.1:8001 restart all"
+    "cd /opt/numenta/products &&
+     git reset --hard ${COMMIT_SHA} &&
```
This seems redundant with line 130

oxtopus replied:
Thanks.  I meant to remove it here.

jcasner commented on taurus/pipeline/scripts/overrides/taurus/conf/nginx-taurus.conf:
```diff
@@ -0,0 +1,80 @@
+    #
+    # See http://wiki.nginx.org/HttpCoreModule#location
+
+    location ^~ /_annotations {
```
Do we need the `_annotations` end point in Taurus?

oxtopus replied:
Probably not, but this is how the nginx config is in the source.  The reason it's duplicated here is I need to change the user from root, to ec2-user.  I support removing the unneeded location directive, but it's outside the scope of this issue in particular.

jcasner commented on taurus/pipeline/scripts/overrides/taurus/conf/ssl/localhost.key:
```diff
@@ -0,0 +1,15 @@
+-----BEGIN RSA PRIVATE KEY-----
```
What is this private key?  For SSH communication between collector and server?

oxtopus replied:
It's the result of following instructions in README for generating self-signed ssl cert.  In practice, this should be discarded for a new set of ssl credentials, but for now is there for batteries-included purposes, since the generating step requires manual intervention (enter a password) and can't be automated.

jcasner replied:
We can update the generation script to pass the password as a command line arg

oxtopus replied:
I may need some help with that.  I wasn't able to get it to work.

jcasner commented:
@oxtopus I'm finished with this round of review.

unixorn commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -20,54 +20,217 @@
+pair of Taurus instances suitable for use and/or testing.
+"
+
+if [ "${1}" == "-h" ]; then
```
Might as well make `--help` work here too

oxtopus commented:
With the exception of the ssl cert, I've addressed your feedback.  I've restored the rabbitmq host var to read from the env as it should, but it looks like the connectivity issue I was facing had to do with the default guest:guest credentials not accepted on a public interface: https://www.rabbitmq.com/access-control.html

jcasner commented on taurus/pipeline/scripts/initialize-remote-taurus-servers.sh:
```diff
@@ -35,39 +123,175 @@ REPOPATH=`dirname "${SCRIPT}"`/../../..
+    openssl genrsa \
+      -des3 \
+      -out taurus/pipeline/scripts/overrides/taurus/conf/ssl/${TAURUS_SERVER_HOST}.key \
+      -passout pass:myvoiceismypassportverifyme \
```
Might be good to tokenize this, despite the excellent reference!

oxtopus replied:
See latest, I made it random.

jcasner commented:
:+1:
